【Feature】消費カロリー手動入力機能の実装 - MVP

masa5281 opened the pull request:
## 概要
本PRでは、消費カロリーと運動日を手動で入力する機能を作成しました。
あわせてヘッダーも作成しております。

また、Tailwind CSSを用いて簡易的にスタイリングを行いました。
## 主な変更点
- **消費カロリー手動入力の保存処理**
- **運動日入力の保存処理**
- **バリデーションの実装**
  - 空文字の禁止
  - 0より大きい値で入力
  - 未来の日付を選択不可
- **ヘッダーを作成**
- **Tailwind CSSを用いて最低限のスタイリング**
- **コンポーネント化**
## 関連するIssues
Close #5 
※詳細なタスクについてはissueをご確認ください。
## レビュワーへのお願い
- 手動入力した消費カロリー、運動日の保存が適切に処理されているかご確認ください
- バリデーションが適切に機能するかご確認をお願いします
- 冗長または不要な記述が無いかご確認いただけると助かります

masa5281 committed:
chore: Reactの初期セットアップファイルを整理

masa5281 committed:
fix: WebSocket接続エラーの解消

- `docker-compose.yml`にコンテナ側のReact開発サーバーのポートを3001に設定
- `package.json`に起動ポート番号を設定

masa5281 committed:
feat: yarnで`axios`を導入（#5）

- RailsとHTTPを用いてAPI通信をするため

masa5281 committed:
feat: Caloriesコントローラーの作成（#5）

masa5281 committed:
feat: CalorieモデルとCaloriesテーブルを作成（#5）

- マイグレーションファイルに、`burned_calorie`・`recorded_at`カラムを定義

masa5281 committed:
feat: カロリー入力保存処理のルーティングを定義（#5）

- resourcesでcreateアクションへのルートを設定

masa5281 committed:
feat: 消費カロリー入力処理のマークアップを実装（#5）

masa5281 committed:
feat: Gem「pry-byebug」、「pry-rails」の追加

masa5281 committed:
fix: gem 'ostruct'で発生していたエラーを修正

- `ostruct`がRubyのデフォルトgemでなくなったことが原因

masa5281 committed:
fix: `calorie`を不規則変化の単語として登録（#5）

- ストロングパラメーターでキーの値が`calory`に推測される問題を修正

masa5281 committed:
feat: 消費カロリー入力機能の保存処理をAPI経由で実装（#5）

- Caloriesコントローラーにcreateアクション、ストロングパラメーター`calorie_params`を実装
- App.jsxにaxiosを用いたRailsへのPOST処理を実装

masa5281 committed:
feat: 消費カロリー入力の保存処理のバリデーションの追加（#5）

- Calorieモデル: `burned_calorie`カラムに`presence`と`numericality`を追加
- Caloriesコントローラー: バリデーションエラーメッセージと422ステータスコードを返す処理を追加
- App.jsx: バリデーションエラーメッセージ表示処理を追加

masa5281 committed:
feat: バリデーションエラーメッセージの日本語化（#5）

- Gemfile: gem`rails-i18n`を追加
- application.rb: デフォルトの言語設定を日本語に設定
- ja.yml: モデル名とカラムに対応した日本語訳を追加

masa5281 committed:
feat: 登録日入力フォームのマークアップを実装（#5）

- ライブラリ: react-datepickerを追加

masa5281 committed:
feat: 登録日入力フォームの保存処理を実装（#5）

- `date-fns`で日本語ロケールを設定
- `toDateString()`を用いてRailsに送信する日付をフォーマット

masa5281 committed:
feat: 登録日入力フォームのバリデーションの追加（#5）

- 未来の日付を指定できないカスタムバリデーションを追加

masa5281 committed:
feat: `PostCSS`と`autoprefixer`の導入

- `PostCSS`: CSSソースコードを変換・最適化するため追加
- `autoprefixer`: ベンダープレフィックスを付与し異なるブラウザ間での差異をなくすため追加

masa5281 committed:
feat: Tailwind CSSのカラーパレットの拡張（#5）

- primary、secondary、background、header、textの各色を設定

masa5281 committed:
feat: React Iconsの追加（#5）

masa5281 committed:
feat: Tailwind CSSで最低限のスタイリングを実装（#5）

- tailwind.config.js: hover時、エラー時のカラーパレットを追加

masa5281 committed:
feat: バリデーションエラーメッセージをカスタマイズ（#5）

- エラーメッセージメッセージを格納したオブジェクトを返すよう実装
- ja.yml: カラムとバリデーションに対応したエラーメッセージを追記
- application.rb: カスタムエラーメッセージを有効化

masa5281 committed:
refactor: named exportに変更

masa5281 committed:
chore: 文言の修正と、スピンボタンの無効化（#5）

masa5281 committed:
feat: アイコン画像の追加（#5）

masa5281 committed:
feat: Tailwind CSSのカラーパレットでshadowの色を拡張（#5）

masa5281 committed:
feat: 基本のヘッダーを作成（#5）

- Tailwind CSSやアイコンを用いて基本的なスタイリングの実装
- リンク先や目標ゲージなどの拡張機能は今後実装予定

masa5281 committed:
fix: 登録日のバリデーションが当日9時まででも通るように修正（#5）

- Railsで設定したタイムゾーン（Tokyo）で当日の日付を正しく取得するよう設定

masa5281 committed:
fix: 消費カロリーの入力で全角→半角変換を実装（#5）

- inputタグのtype属性をtext型に変更
- onChangeToText関数で全角入力を半角入力へ変換する処理を追記

masa5281 committed:
refactor: アイコン画像をpublic/imagesからsrc/imagesへ移動

masa5281 committed:
refactor: ヘッダーをコンポーネント化（#5）

- App.jsxからHeaderコンポーネントをインポートし、再利用可能に変更

masa5281 committed:
refactor: カロリー入力フォームの日付入力をコンポーネント化（#5）

masa5281 committed:
refactor: バリデーションエラーメッセージをコンポーネント化（#5）

masa5281 committed:
refactor: カロリー入力フォームの換算ボタンをコンポーネント化（#5）

masa5281 committed:
refactor: カロリー入力フォームのコンポーネント化（手動部分のみ）（#5）

- CalorieInput: カロリー入力画面を表示する
- ManualCalorieForm: カロリー入力フォーム（手動）を表示する

※日付入力、エラーメッセージ、換算ボタンのコンポーネントは別コミットで管理済み

masa5281 committed:
refactor: Tailwind CSSの書き順を「視覚順」に整列（#5）

TakeshiAndreOnaga commented on backend/app/controllers/calories_controller.rb:
```diff
@@ -0,0 +1,18 @@
+  
+  private
+  
+  # フロントから送られてくるデータ
```
特殊な書き方をしている形ではないので、他のストロングパラメータメソッド同様にこちらのコメントは不要です！

masa5281 replied:
コメントを削除いたしました！
ご確認をお願いいたします。

TakeshiAndreOnaga commented on frontend/src/components/CalorieInput/ManualCalorieForm.jsx:
```diff
@@ -0,0 +1,61 @@
+
+  const createManualCalorie = async () => {
+    try {
+      await axios.post("http://localhost:3000/calories", {
```
できればこの http://localhost:3000 の部分を環境変数を使用して定義するようにしたいですね。（Dockerを使用して私も動かしたことがないので使えるか怪しいですが...）
環境変数を使用する理由はこれを本番環境にデプロイする時URLが変わる可能性が高いためです。事前にdevelopment環境ではlocalhost、本番環境ではxxxのようにしておくとデプロイした時にページ遷移エラーがなくなります。

参考になるかもです！↓
https://ralacode.com/blog/post/use-env-variables-in-react/

masa5281 replied:
Rails、React両方で環境変数を参照するよう修正いたしました。
デプロイが近づいてきた際に`.env.production`ファイルに本番用のURLを記載したいと思います！

ご確認をお願いいたします。

masa5281 committed:
refactor: ストロングパラメータの不要なコミットを削除（#6）

masa5281 committed:
refactor: 環境変数でフロントとバックエンドURLを管理（#6）

- api.js: axiosのインスタンスを設定
- Gemfile: 環境変数を管理するため`dotenv-rails`を導入
- docker-compose.yml: .envファイルを有効化
